### ISSUE TITLE
Please consider writing to target directory directly, i.e., no tmp file

### ISSUE THREAD

**Issue by shuangye** [Densaugeo/uploadserver#22]
I want to upload some large files (size in GB) to an external USB drive. I found that during file uploading, the external USB drive LED does not flash. The computer internal LED flashes instead.
I opened the macOS Activity Monitor, and selected the Python process. In "Open Files and Ports", I found that the uploaded file was saved as a tmp file in computer internal disk, like this 
`/private/var/folders/kr/pn0vp9sx703fhy8v8d6fm8b80000gn/T/tmpfcqc7li5`
and then wrote to the external USB drive.
That is, one upload generates 2 file copies. It's OK for few small files, but it takes more time when uploading large files, and extra writing decreases the lifetime of computer internal disk (SSD).

I want to save the file directly to the external USB drive, without tmp files. Could you please consider this improvement? Thank you.

macOS 12.4 Monterey. Python 3.8.9
Intel i7 CPU, 32GB DDRAM.

**Comment by Densaugeo** [comment]
Currently, caching files during upload is handled by the underlying `http.server` module from Python's built-in libraries. There have been some suggestions to switch to streaming files directly to disk during upload, but so far all of these have required new dependencies that I want to avoid. Much of what makes `uploadserver` simple and lightweight is that it uses nothing outside Python's built-in modules.

If you do find a way to stream files from `http.server` without requiring additional dependencies or introducing much complexity, please reopen the issue.

**Comment by shuangye** [comment]
I found a solution. The idea is to change tmp path to upload dir. Rename the uploaded file, instead of copying it.
This will save the life of your computer's SSD.

1. In main():
`tempfile.tempdir = args.directory`

2. Derive the cgi.FieldStorage class:
```
class PersistentFieldStorage(cgi.FieldStorage):
    def make_file(self):
        if self._binary_file:
            return tempfile.NamedTemporaryFile(mode='wb+', delete = False)
        else:
            return tempfile.NamedTemporaryFile("w+", delete = False,
                encoding=self.encoding, newline = '\n')
```

3. Rename the uploaded file in receive_upload():
```
if filename:
    dst = pathlib.Path(args.directory) / filename
    field.file.close()
    os.rename(field.file.name, dst)
    handler.log_message('Upload of "{}" accepted as {}'.format(filename, dst))
    result = (http.HTTPStatus.NO_CONTENT, None)
```

**Comment by shuangye** [comment]
cgi.py does not always create the tmp file. It will only create when file size > 1000. See https://github.com/python/cpython/blob/feb4be1553c947ddb2bd8307419451bfdfc0a6a3/Lib/cgi.py#L717
`if self.__file.tell() + len(line) > 1000:`
So the code above need to be refined.

**Comment by Densaugeo** [comment]
Interesting, I'll look into this further when I get some time. A few issues to check:
- Does changing the temp setting affect other `http.server` instances?
- Needs testing on Windows.
- I'm pretty sure there's a pre-existing DoS vulnerability related to uploading files, since authentication is not checked until after a file is received. Maybe if we're tinkering with the upload cache we could address that issue too?

In any case, this looks like it might help with some RAM use issues related to large file uploads as well.

**Comment by shuangye** [comment]
Hello,
- `tempfile.tempdir = args.directory` has global impact to all temp files. An alternative is to use `tempfile.NamedTemporaryFile(mode = 'wb+', dir = args.directory, delete = False)`, wich impacts that call only.
- Tested on:
  - Ubuntu, `strace -o strace.log -ff python3 uploadserver.py 8080 -d ~/` and confirmed that only ONE file generated after uploading.
  - Windows 10, using Procmon64 from [SysInternals Suite](https://learn.microsoft.com/en-us/sysinternals/downloads/sysinternals-suite) , and confirmed that only ONE file generated after uploading.
- I noticed that too. Maybe you can send the token first, using JavaScript. But I do not know how to implement that. I'm not a web developer.

**Comment by Densaugeo** [comment]
Apologies for the delay getting back to you, there was a hurricane and some power outages...

Overall, this looks like an excellent improvement. Memory use testing shows that it can handle large file uploads without exceeding its usual ~16MB RAM footprint.

A few comments on your PR:
- `PersistentFieldStorage` may need to be updated if `cgi.FieldStorage` changes. Please include a comment noting that `make_file` is a modification of the `make_file` function in `cgi.FieldStorage`, and the Python version used, so I'll remember what you did here if it needs updating in the future.
- Please remove `auto_rename` for now. This would be a breaking change, because it would prevent using `uploadserver` to directly replace files. You could make another PR for this change and we could discuss putting it behind a command line option.
- Please use full words for variable names (such as `destination` instead of `dst`).
- The addition to the file progress monitor is nice! Please change `KB` to `kiB`, in keeping with metric standards (since a factor of 1024 is prefixed by `ki`, but `k` is for a factor of 1000).

For test cases, I think setting up a test system to check RAM use and temporary file creation would be more trouble than it is worth, so I'll skip that for now. I'll add a test case with a >1 kB to make sure both code paths for saving the file are covered.

Send an update once you've made the changes and I'll get started on making a new release!

**Comment by shuangye** [comment]
Thank you for your comments. I have pushed an update.
Sorry my editor auto removes trailing whitespaces, so you will see lots of changes...

**Comment by shuangye** [comment]
And pushed another commit with auto-renaming option. Will create a pull request once the current one is merged.

**Comment by Densaugeo** [comment]
Both of your commits are in the same PR, so they would be merged together. Please separate them into different PRs so I can merge them separately.

When we add the auto-rename option, I will have a couple comments that as well: avoid defining functions inside other functions, and it's worth having a discussion about behavior - we may want to make the new auto-renaming the default behavior for security reasons, or perhaps keep the old behavior as default for compatibility. We can discuss that in a new thread when you make another PR.

In any case, once you have separated the auto-rename option to another PR and straightened out your trailing spaces issue with your text editor, I can merge the target directory change. (Not sure what editor you're using, but most editors that modify white space have an option to disable that).

**Comment by shuangye** [comment]
Although having trailing whitespaces is not allowed in many coding  conventions, I put them back, to make the commit neat.

**Comment by Densaugeo** [comment]
Merged and added a test case for the >1 kiB code path.

Thanks for sending in this patch!

I'll wait on making a new release until we merge the auto-rename update.
